### PR TITLE
first update

### PR DESCRIPTION
**Pull request by fedeimbaguable**

### CODE REVIEW

File: pages/CheckingPage.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import BasePage from '../pages/BasePage';
+
+class CheckingPage extends BasePage {
```
Comment (rquintana-abstracta) [inline pages/CheckingPage.js]: This class mixes elements from the checking account creation page and the one about viewing existing accounts. Make sure to split this into two classes and allocate elements accordingly. This will make this class more focused (SRP adherent: https://en.wikipedia.org/wiki/Single-responsibility_principle)

File: pages/CheckingPage.js [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    get newCheckingFormResetButton () {return $('#newCheckingReset')}
+
+    get newCheckingConfirmation () {return $('#new-account-conf-alert')}
+
+    get newCheckingErrorText () {return $('#new-account-error-msg')}
+
+    get switchChecking () { return $('.switch-label')}
+
+    get checkingBody () { return $('.even')}
```
Comment (rquintana-abstracta) [inline pages/CheckingPage.js]: This is not a robust selector. See if you can craft one that better individualizes this element. I assume this is the "card" that represents the account after account creation.

File: pages/CheckingPage.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    get checkingFormNextButton () {return $('#transactionTable_next')}
```
Comment (rquintana-abstracta) [inline pages/CheckingPage.js]: Is this referring to the next and previous buttons from the trasnactions table (as ids suggests)? If so, this name is misleading. Re-name it to `checkingTransactionsTableNextButton`

File: pages/CheckingPage.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    get checkingFormPreviousButton () {return $('#transactionTable_previous')}
```
Comment (rquintana-abstracta) [inline pages/CheckingPage.js]: Same comment as above.

File: pages/CheckingPage.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    get checkingTableTextInfo(){return $('#transactionTable_info')}
+
+    get checkingSearchInput(){ return $('[type="search"]') }
+
+    async changeCheckingFormPageNext(){
```
Comment (rquintana-abstracta) [inline pages/CheckingPage.js]: This is the transactions table, not the checking form. Rename it accordingly

File: pages/CheckingPage.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        await this.checkingFormNextButton.click()
+    }
+    async changeCheckingFormPagePrevious(){
```
Comment (rquintana-abstracta) [inline pages/CheckingPage.js]: Same comment about making the name refer to the trasnactions table instead of the form

File: pages/CheckingPage.js [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+  
+    get newCheckingFormSubmitButton () {return $('#newCheckingSubmit')}
+
+    get newCheckingFormResetButton () {return $('#newCheckingReset')}
+
+    get newCheckingConfirmation () {return $('#new-account-conf-alert')}
+
+    get newCheckingErrorText () {return $('#new-account-error-msg')}
+
+    get switchChecking () { return $('.switch-label')}
```
Comment (rquintana-abstracta) [inline pages/CheckingPage.js]: Every selector that points to something on the "form/card" that represents the summary of a specific checking account has to reference to the specific account it wants, as that representation is actually part of a collection of accounts that renders one per existing account (therefore, as you create accounts, you'll have "n" replicas of the form/card, so this class will be present in all of them).

Make sure to consider that scenario in your solution (crafting a selector that points to a specific account)

File: tests/checkings.test.js [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+import SignIn from "../pages/SignIn";
+import CheckingPage from "../pages/CheckingPage"
+ 
+
+describe('Checkings', () => {
+it('Should take you to Checkings form', async ()=> {
+    await SignIn.open('bank')
+    let username = 'jsmith@demo.io'
+    let password = 'Demo123!'
+    await SignIn.logIn(username, password)
```
Comment (rquintana-abstracta) [inline tests/checkings.test.js]: Consider you'll need to login for every test in this suite, not just the first one. If not, you'll make them dependent of the order they'll be defined (which is an antipattern for tests).

Test frameworks provide a clean abstraction for dealing with this scneario (executing a piece of logic for every test on a suite) called **hooks** (in particular, the hook that fits this is the **beforeEach** hook).

Implement this login logic for this suite in a beforeEach hook

Reference: https://mochajs.org/#hooks

BTW, wdio uses lots of already existent technologies under the hood: **Mocha** is one of them, for everything having to do with test definition / execution, **Jest** for matchers/asserts, so anything you read there is valid for wdio as well.

File: pages/CheckingPage.js (hunk 2 above)
Comment (fedeimbaguable) [reply pages/CheckingPage.js]: I dont use it as a selector, as tr in the table are named by odd and even, when I create a Checking that card is selected so when I do the switch to the one Digital Bank provides me if a tr is even appears that means I change the checking view because the checking I created has only one transaction

File: pages/CheckingPage.js (hunk 7 above)
Comment (fedeimbaguable) [reply pages/CheckingPage.js]: I use the switch checking only when two cards are created, so by pressing the switch label it changes the card, the value of each card changes when I do the tests because I have to delete and create data.

File: pages/HomePage.js [hunk 9]
```diff
@@ -7,8 +7,28 @@ class HomePage extends BasePage {
...
   get userAvatar() { return $('img[alt="User Avatar"]')};
 
-  get logOutButton() {return $('a[href="/bank/logout"]')}
+  get logOutButton() {return $('a[href="/bank/logout"]')};
 
+  get newCheckingForm () {return $('#new-checking-menu-item')}
+
+  get checkingMenu () {return $('#checking-menu')}
+
+  get checkingTitle () { return $('#page-title') }
```
Comment (rquintana-abstracta) [inline pages/HomePage.js]: This is part of the checkings page, locate it there.

File: tests/checkings.test.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    await CheckingPage.changeCheckingFormPageNext()
+    await expect(CheckingPage.checkingTableTextInfo).toHaveText('Showing 21 to 24 of 24 entries')
+    await CheckingPage.changeCheckingFormPagePrevious()
+    await expect(CheckingPage.checkingTableTextInfo).toHaveText('Showing 11 to 20 of 24 entries')
+    await CheckingPage.changeCheckingFormPagePrevious()
+    await expect(CheckingPage.checkingTableTextInfo).toHaveText('Showing 1 to 10 of 24 entries')
+})
+it('Should filter the table with the word "Income"', async ()=> {
+    await CheckingPage.sendText(CheckingPage.checkingSearchInput, 'Income')
+    await expect(CheckingPage.checkingTableTextInfo).toHaveTextContaining('(filtered from 24 total entries)')
```
Comment (rquintana-abstracta) [inline tests/checkings.test.js]: IMO, in order to assert this scenario, you have to make sure no other entry but the ones that have text "Income" are displayed in the table after the search. Therefore you need to implement a process that checks every row and validates that all of them (in the relevant column) displays "Income". If one of them don't, then the test fails.

Implement the assertion considering that logic

File: tests/checkings.test.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+})
+it('Should reset the data', async ()=> {
```
Comment (rquintana-abstracta) [inline tests/checkings.test.js]: Maybe describe this a little bit more to make it more expicit. Something like: "Should reset the account creation form when clicking reset"

File: tests/checkings.test.js [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    await HomePage.displayCheckingMenu() 
+    await HomePage.accessNewCheckingForm()
+    await CheckingPage.sendText(CheckingPage.checkingNameInput, '1234')
+    await CheckingPage.checkingReset()
+    await expect(CheckingPage.checkingNameInput).toHaveText('')
+})
+it('Should create a Standard Individual Checking', async ()=> {
+    await HomePage.displayCheckingMenu()
```
Comment (rquintana-abstracta) [inline tests/checkings.test.js:50]: Every test about creating an account shares the same overall structure, only changing the parameters sent for the inputs and (maybe a bit more subtle) the type of accounts radiobuttons options. 

See if you can somehow parameterize the selection of account type with a method (maybe something that takes the options for the account as parameters, such as `createAccount(type, ownership, name, amount)`, **type** being a string that could be standard or interest, and **ownership** being if its individual or joint), and then use a **data driven** approach to only define it once, and then run it with every combination of parameters possible.

File: tests/checkings.test.js [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+})
+it('Should not create a checking because the deposit is less than 25', async ()=> {
+    await HomePage.displayCheckingMenu() 
+    await HomePage.accessNewCheckingForm()
+    await CheckingPage.chooseStandardChecking()
+    await CheckingPage.chooseIndividualChecking()
+    await CheckingPage.sendText(CheckingPage.checkingNameInput, 'Error')
+    await CheckingPage.sendText(CheckingPage.checkingDepositInput, '24.99')
+    await CheckingPage.checkingSubmit()
+    await expect(CheckingPage.newCheckingErrorText).toBeDisplayed()
```
Comment (rquintana-abstracta) [inline tests/checkings.test.js]: You can try more types of invalid input and make tests for it (maybe using a **data driven** approach as well), like negative numbers, characters, empty, etc

File: pages/CheckingPage.js (hunk 2 above)
Comment (rquintana-abstracta) [reply pages/CheckingPage.js]: I understand. This is, IMO, too reliant on the specific data is present in the environment at the moment you run the test. If you ever make a new transaction to the account you just created it'll be selected instead, breaking your test.

I think a more robust approach would be relying on the value of the switch after you activate it (it should only display one as "checked", the rest as "unchecked"). This approach won't be validating that the table actually displays different data, but at least it's not bounded to specific data being loaded on the environment for it to work properly.

File: pages/CheckingPage.js (hunk 7 above)
Comment (rquintana-abstracta) [reply pages/CheckingPage.js]: What I mean is that this class (switch-label) is present on each "card" representing an account, so it doesn't individualize a specific account, it'll grab the first one it encounters with that class. This binds it to the specific order structure the cards have. That's what makes it less robust. You can craft a selector that uniquely identifies a specific account that you want, independent of order, only based on account's attributes, using **xpath**. 

You can take a look at this topic for inspiration, where we solved a similar problem with xpath. Ref: https://abstracta.ryver.com/#posts/2957788

File: tests/checkings.test.js (hunk 12 above)
Comment (fedeimbaguable) [reply tests/checkings.test.js:50]: How do I transform the string into the element it has to click?

**Review by rquintana-abstracta — CHANGES_REQUESTED** [review]
Implement requested changes

File: tests/checkings.test.js (hunk 12 above)
Comment (rquintana-abstracta) [reply tests/checkings.test.js:50]: Use an if statement or ternary operator to implement it (preferably ternary operator for readability/shortness). Something like:
```javascript
string == "individual" ? await this.selectIndividualAccount() : await this.selectJointAccount()
```

BTW, I don't remember the exact method names you used for those methods, but this is the idea. If string is "individual", it executes first part of the expression, else the other part: Ref: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator

File: pages/CheckingPage.js (hunk 2 above)
Comment (fedeimbaguable) [reply pages/CheckingPage.js]: The values in the switch dont change when they are off or on

File: pages/CheckingPage.js (hunk 7 above)
Comment (fedeimbaguable) [reply pages/CheckingPage.js]: How does the tree works when I want to get it as an element?

File: tests/checkings.test.js (hunk 10 above)
Comment (fedeimbaguable) [reply tests/checkings.test.js]: By using a class, does the test checks every class or only the first one?

File: tests/checkings.test.js (hunk 13 above)
Comment (fedeimbaguable) [reply tests/checkings.test.js]: I dont know how to assert it because a pop up appears.

Comment (rquintana-abstracta) [reply tests/checkings.test.js]: Assert that you stay in the same page instead of being redirected to the checkings-view, for example, pretty much signaling that the account was not created.

File: tests/checkings.test.js [hunk 14]
```diff
@@ -1,96 +1,75 @@
...
+import CheckingPageForm from "../pages/CheckinPageForm"
+import CheckingPageTable from "../pages/CheckingPageTable";
+import CorrectCheckings from "../datos/CorrectCheckings";
+import IncorrectCheckings from "../datos/IncorrectCheckings"
  
-
-describe('Checkings', () => {
-it('Should take you to Checkings form', async ()=> {
-    await SignIn.open('bank')
+beforeEach(function () {
```
Comment (rquintana-abstracta) [inline tests/checkings.test.js]: You need to make this function ayncrhonous, as you are performing asynchronous operations inside it. Turn the callback function itno an async function and add `await` on each line that interacts with web elements somehow, if not, this part is not going to be executed adequately.

```javascript
beforeEach(async () => {
// logic with pontentially await somewhere goes here
}
```

File: tests/checkings.test.js [hunk 15]
```diff
@@ -1,96 +1,75 @@
...
-    await CheckingPage.chooseIndividualChecking()
-    await CheckingPage.sendText(CheckingPage.checkingNameInput, 'Interest Individual Checking')
-    await CheckingPage.sendText(CheckingPage.checkingDepositInput, '1000000')
-    await CheckingPage.checkingSubmit()
-    await expect(CheckingPage.newCheckingConfirmation).toBeDisplayed()
+    await HomePage.accessViewCheckingTable()
+    await CheckingPageTable.sendText(CheckingPageTable.checkingSearchInput, 'Income')
+    const rows = await $$('#transactionTable tr');
+    const columns = await rows[1,2,3,4].$$('td');
+    await expect(columns[1]).toHaveText('Income');
```
Comment (rquintana-abstracta) [inline tests/checkings.test.js]: Remember about keeping abstraction layers separated, so no direct selectors should be used at test layer, but always interacted through page objects. Trun this into methods on the corresponding page object and implement this assertion considering that structure.

Also, the general approach you are using is basically correct (reading rows as an array of rows, and then checking columns looking for instances of the "Income" text), but there are flaws in the implementation. Once you have the rows, you need to extract a specific column from each of them, one at a time. `rows[1,2,3,4]` is not valid syntax here, and this will break if you run it (make sure to run tests after changes to make sure they keep working).

One simple approach for solving this would be using a parameterized css selector, that returns you the relevant column from the provided row, such as: 

```javascript

`#transactionTable>tbody>tr:nth-child(${row})>td:nth-child(4)`

```

Where `row` in this case is passed as parameter to the method that takes care of extracting this. This selector will return the element representing the "cell" belonging to the 4th column from the nth row (that row number being controlled by the parameter).

After having that, you can do (in the test), something like:

```javascript

let tableLength = await CheckingPage.getTableNumberOfRows();
let cell;
for(let i=0; i< tableLength; i++){
     cell = await CheckingPage.getCellFromForRow(i);
     await expect(cell).toHaveTextContaining("Income");
}
```

Or something like that. Notice you'll have to implement a couple of auxiliary methods for this

File: tests/checkings.test.js [hunk 16]
```diff
@@ -1,96 +1,75 @@
...
+CorrectCheckings.forEach(({type, ownership, name, amount, reason}) => {
+    it(`Should create a  ${reason}`, async () => {
+        await HomePage.displayCheckingMenu() 
+        await HomePage.accessNewCheckingForm()
+        await CheckingPageForm.createChecking(type, ownership, name, amount)
+        await expect(CheckingPageTable.newCheckingConfirmation).toBeDisplayed()
+    })
+    });
+
+//it('Should switch checking', async ()=> {
```
Comment (rquintana-abstracta) [inline tests/checkings.test.js]: Delete commented code

**Review by rquintana-abstracta — CHANGES_REQUESTED** [review]
Implement new changes

File: pages/CheckingPage.js (hunk 2 above)
Comment (rquintana-abstracta) [reply pages/CheckingPage.js]: The get assigned a "checked" attribute when they are checked, and that attribute doesn't appear when it is unchecked. You can use that as a signal

File: tests/checkings.test.js (hunk 15 above)
Comment (fedeimbaguable) [reply tests/checkings.test.js]: Row, is an array of the table length? How do I get the number of rows in the DOM?

Comment (rquintana-abstracta) [reply tests/checkings.test.js]: `row` in the example selector I provided is an index, a numerical value representing the order of the row you want to retrieve.

For the second action, you can locate all rows from the transactions table and parse them into an array (using `$$('tr')` syntax) and then return the array's length, or something like that

File: pages/CheckingPageTable.js [hunk 17]
```diff
@@ -4,17 +4,19 @@ class CheckingPageTable extends BasePage {
...
     get checkingTransactionsTableNextButton () {return $('#transactionTable_next')}
 
     get checkingTransactionsTablePreviousButton () {return $('#transactionTable_previous')}
 
     get checkingTableTextInfo(){return $('#transactionTable_info')}
 
-    get checkingTableFilter(){return $('tbody')}
+    get getCellFromForRow(){return $(`#transactionTable>tbody>tr:nth-child(${row})>td:nth-child(2)`)}
+
+    get getTableNumberOfRows(){return $('tbody')}
```
Comment (rquintana-abstracta) [inline pages/CheckingPageTable.js]: This is returning a web element representing the whole table. It's not returning the number of rows the table has. Use a selector that matches the rows of the table, and use array based syntax to get them into an array of row elements. Then just return its length, like:

```javascript
let rows = await $$("tbody > tr");
return rows.length
```

This is not a getter method, its a regular async method. Implement it without the `get` modifier

Reference: https://webdriver.io/docs/api/browser/$$/

File: pages/CheckingPageTable.js [hunk 18]
```diff
@@ -4,17 +4,19 @@ class CheckingPageTable extends BasePage {
...
+    get switchChecked () { return $('//*[contains(text(),"Individual Checking")]//ancestor::div[@class="card-body"]//input[@class="switch-input"]')}
 
     get checkingTransactionsTableNextButton () {return $('#transactionTable_next')}
 
     get checkingTransactionsTablePreviousButton () {return $('#transactionTable_previous')}
 
     get checkingTableTextInfo(){return $('#transactionTable_info')}
 
-    get checkingTableFilter(){return $('tbody')}
+    get getCellFromForRow(){return $(`#transactionTable>tbody>tr:nth-child(${row})>td:nth-child(2)`)}
```
Comment (rquintana-abstracta) [inline pages/CheckingPageTable.js]: You're not passing this the row parameter that the body of this method is requesting to build the selector string. Also, this is a regular async method as well, not a getter method. Transform it accordingly.

Also, this `fromFor` seems to be a typo from when I sent it this snippet as example, its just `getCellFromRow()`

**Review by rquintana-abstracta — CHANGES_REQUESTED** [review]
Implement requested changes

**Review by rquintana-abstracta — APPROVED** [review]
